File: src/services/api/dialog.ts
/**
 * 系统对话框能力封装层（services/api）。
 *
 * 作为渲染层访问 Tauri 原生对话框系统能力的统一入口，
 * 负责打开文件、另存为、覆盖/关闭确认弹窗以及通用的文件/目录选择对话框。
 */
import type { FileTraits, ReadFileResult, SaveFileResult } from "@/shared/types/export";
import { invoke } from "@/composables/useTauri";

/**
 * 弹出系统"打开文件"对话框，让用户选择一个 Markdown 文件并读取其内容。
 *
 * @returns 读取结果（含内容与文件特征）；用户取消时返回 null
 * @remarks 调用 Tauri command `open_file`
 */
export async function openFile(): Promise<ReadFileResult | null> {
  return invoke<ReadFileResult | null>("open_file");
}

/** "另存为"对话框的入参。 */
export interface SaveFileAsArgs {
  /** 待保存的文件文本内容 */
  content: string;
  /** 源文件特征（编码、换行符等），用于保持一致性 */
  fileTraits?: FileTraits;
  /** 无源文件特征时使用的默认换行符 */
  defaultLineEnding?: FileTraits["lineEnding"];
  /** 本地图片存放路径，用于迁移内容引用的本地图片 */
  imageLocalPath?: string;
  /** 对话框默认打开/定位的路径 */
  defaultPath?: string;
  /** 对话框默认填充的文件名 */
  fileName?: string;
}

/**
 * 弹出系统"另存为"对话框，将内容写入用户选择的新位置。
 *
 * @param args 另存为参数，见 {@link SaveFileAsArgs}
 * @returns 保存结果（含最终路径）；用户取消时返回 null
 * @remarks 调用 Tauri command `save_file_as`
 */
export async function saveFileAs(args: SaveFileAsArgs): Promise<SaveFileResult | null> {
  return invoke<SaveFileResult | null>("save_file_as", { args });
}

/** 覆盖确认对话框的返回选项：0=取消 1=覆盖 2=另存为。 */
export type OverwriteChoice = 0 | 1 | 2; // 0=取消 1=覆盖 2=保存
/**
 * 弹出"文件已存在，是否覆盖"确认对话框。
 *
 * @param fileName 冲突的文件名，用于提示文案
 * @returns 用户选择，见 {@link OverwriteChoice}
 * @remarks 调用 Tauri command `show_overwrite_confirm`
 */
export async function showOverwriteConfirm(fileName: string): Promise<OverwriteChoice> {
  return invoke<OverwriteChoice>("show_overwrite_confirm", { fileName });
}

/** 关闭确认对话框的返回选项：0=取消 1=不保存 2=保存。 */
export type CloseChoice = 0 | 1 | 2; // 0=取消 1=不保存 2=保存
/**
 * 弹出"文件未保存，关闭前是否保存"确认对话框。
 *
 * @param fileName 待关闭文件的文件名，用于提示文案
 * @returns 用户选择，见 {@link CloseChoice}
 * @remarks 调用 Tauri command `show_close_confirm`
 */
export async function showCloseConfirm(fileName: string): Promise<CloseChoice> {
  return invoke<CloseChoice>("show_close_confirm", { fileName });
}

/** 通用"打开"对话框的配置项。 */
export interface OpenDialogOptions {
  /** 对话框标题 */
  title?: string;
  /** 默认定位路径 */
  defaultPath?: string;
  /** 文件类型过滤器（名称 + 扩展名列表） */
  filters?: Array<{ name: string; extensions: string[] }>;
  /** 是否选择目录而非文件 */
  directory?: boolean;
  /** 是否允许多选 */
  multiple?: boolean;
}

/** 通用"打开"对话框的返回结果。 */
export interface OpenDialogResult {
  /** 用户是否取消了选择 */
  canceled: boolean;
  /** 选中的文件/目录绝对路径列表 */
  filePaths: string[];
}

/**
 * 弹出通用的"打开文件/目录"对话框。
 *
 * @param options 对话框配置，见 {@link OpenDialogOptions}
 * @returns 选择结果，见 {@link OpenDialogResult}
 * @remarks 调用 Tauri command `show_open_dialog`
 */
export async function showOpenDialog(options: OpenDialogOptions): Promise<OpenDialogResult> {
  return invoke<OpenDialogResult>("show_open_dialog", { options });
}

File: src-tauri/src/commands/dialog.rs
//! 对话框相关 command。对应 IPC：
//!   - `dialog:openFile`           → open_file
//!   - `dialog:saveFileAs`         → save_file_as
//!   - `dialog:showOverwriteConfirm` → show_overwrite_confirm
//!   - `dialog:showCloseConfirm`   → show_close_confirm
//!   - `dialog:showOpenDialog`     → show_open_dialog

use std::path::PathBuf;

use serde::{Deserialize, Serialize};
use tauri::{AppHandle, Manager};
use tauri_plugin_dialog::{DialogExt, FilePath, MessageDialogButtons, MessageDialogKind};

use crate::commands::file::{ReadFileResult, SaveFileResult};
use crate::error::{AppError, AppResult};
use crate::file_format::{FileTraits, LineEnding, restore_file_traits};
use crate::image::prepare_image_content_for_save;
use crate::markdown_file::read_markdown_file;

/// 打开文件对话框并读取选中的 Markdown。取消时返回 None。
#[tauri::command]
pub async fn open_file(app: AppHandle) -> AppResult<Option<ReadFileResult>> {
    let picked = app
        .dialog()
        .file()
        .add_filter("Markdown", &["md", "markdown"])
        .blocking_pick_file();

    let Some(picked) = picked else {
        return Ok(None);
    };
    let file_path = match picked {
        FilePath::Path(p) => p,
        FilePath::Url(u) => PathBuf::from(u.path()),
    };

    let out = tokio::task::spawn_blocking(move || {
        read_markdown_file(&file_path.to_string_lossy())
    })
    .await
    .map_err(|e| anyhow::anyhow!(e))??;

    Ok(out.map(|o| ReadFileResult {
        file_path: o.file_path.to_string_lossy().into_owned(),
        content: o.content,
        file_traits: o.file_traits,
    }))
}

/// `save_file_as` 的入参：内容、原文件格式特征、默认换行风格、图片本地目录、默认保存路径与文件名。
#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SaveFileAsArgs {
    pub content: String,
    pub file_traits: Option<FileTraits>,
    pub default_line_ending: Option<LineEnding>,
    pub image_local_path: Option<String>,
    pub default_path: Option<String>,
    pub file_name: Option<String>,
}

/// 另存为：弹保存对话框 + 写盘 + 临时图片迁移 + traits 还原。
#[tauri::command]
pub async fn save_file_as(
    app: AppHandle,
    args: SaveFileAsArgs,
) -> AppResult<Option<SaveFileResult>> {
    let mut builder = app.dialog().file().add_filter("Markdown", &["md", "markdown"]);
    if let Some(default_path) = &args.default_path {
        builder = builder.set_directory(default_path);
    }
    builder = builder.set_file_name(args.file_name.as_deref().unwrap_or("Untitled.md"));
    let picked = builder.blocking_save_file();

    let Some(picked) = picked else {
        return Ok(None);
    };
    let target = match picked {
        FilePath::Path(p) => p,
        FilePath::Url(u) => PathBuf::from(u.path()),
    };

    let user_data_dir = app
        .path()
        .app_data_dir()
        .map_err(|e| AppError::Other(anyhow::anyhow!("resolve userData dir: {e}")))?;

    let SaveFileAsArgs {
        content,
        file_traits,
        default_line_ending,
        image_local_path,
        default_path: _,
        file_name: _,
    } = args;

    let result: AppResult<SaveFileResult> = tokio::task::spawn_blocking(move || {
        let prepared = prepare_image_content_for_save(
            &content,
            &target,
            image_local_path.as_deref().unwrap_or("/assets"),
            &user_data_dir,
        )?;
        let restored = restore_file_traits(&prepared, file_traits.as_ref(), default_line_ending);
        std::fs::write(&target, restored)?;
        Ok(SaveFileResult {
            file_path: target.to_string_lossy().into_owned(),
            content: prepared,
        })
    })
    .await
    .map_err(|e| anyhow::anyhow!(e))?;

    result.map(Some)
}

/// 覆盖确认：返回 0=取消 1=覆盖 2=保存（与当前桌面端交互约定对齐）。
#[tauri::command]
pub async fn show_overwrite_confirm(app: AppHandle, file_name: String) -> AppResult<u8> {
    let overwrite = app
        .dialog()
        .message(format!(r#"文件 "{file_name}" 已存在，是否要覆盖当前内容？"#))
        .kind(MessageDialogKind::Info)
        .title("文件已存在")
        .buttons(MessageDialogButtons::YesNo)
        .blocking_show();
    if overwrite {
        return Ok(1);
    }
    let save = app
        .dialog()
        .message("是否先保存当前内容后再打开新文件？")
        .kind(MessageDialogKind::Info)
        .title("保存当前内容")
        .buttons(MessageDialogButtons::YesNo)
        .blocking_show();
    if save { Ok(2) } else { Ok(0) }
}

/// 关闭确认：返回 0=取消 1=不保存 2=保存。
#[tauri::command]
pub async fn show_close_confirm(app: AppHandle, file_name: String) -> AppResult<u8> {
    let save = app
        .dialog()
        .message(format!(r#"文件 "{file_name}" 有未保存的更改，是否要保存？"#))
        .kind(MessageDialogKind::Warning)
        .title("文件未保存")
        .buttons(MessageDialogButtons::YesNo)
        .blocking_show();
    if save {
        return Ok(2);
    }
    let discard = app
        .dialog()
        .message("确定不保存吗？未保存的更改将会丢失。")
        .kind(MessageDialogKind::Warning)
        .title("确认不保存")
        .buttons(MessageDialogButtons::YesNo)
        .blocking_show();
    if discard { Ok(1) } else { Ok(0) }
}

/// `show_open_dialog` 的入参：标题、默认路径、文件类型过滤器、是否选目录、是否多选。
#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct OpenDialogOptions {
    pub title: Option<String>,
    pub default_path: Option<String>,
    pub filters: Option<Vec<OpenDialogFilter>>,
    #[serde(default)]
    pub directory: bool,
    #[serde(default)]
    pub multiple: bool,
}

/// 文件类型过滤器：显示名 + 扩展名列表。
#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct OpenDialogFilter {
    pub name: String,
    pub extensions: Vec<String>,
}

/// 打开对话框结果：是否被用户取消，以及选中的路径列表。
#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct OpenDialogResult {
    pub canceled: bool,
    pub file_paths: Vec<String>,
}

/// 通用打开对话框：按 `options` 选择文件或目录、单选或多选，返回选中路径。
///
/// 调用 Tauri dialog 插件的阻塞式选择 API，用户取消时 `canceled = true`。
#[tauri::command]
pub async fn show_open_dialog(
    app: AppHandle,
    options: OpenDialogOptions,
) -> AppResult<OpenDialogResult> {
    let mut builder = app.dialog().file();
    if let Some(t) = &options.title {
        builder = builder.set_title(t);
    }
    if let Some(p) = &options.default_path {
        builder = builder.set_directory(p);
    }
    if let Some(filters) = &options.filters {
        for f in filters {
            let exts: Vec<&str> = f.extensions.iter().map(String::as_str).collect();
            builder = builder.add_filter(&f.name, &exts);
        }
    }

    let paths = if options.directory {
        if options.multiple {
            builder
                .blocking_pick_folders()
                .map(|v| v.into_iter().map(file_path_to_string).collect())
        } else {
            builder
                .blocking_pick_folder()
                .map(|fp| vec![file_path_to_string(fp)])
        }
    } else if options.multiple {
        builder
            .blocking_pick_files()
            .map(|v| v.into_iter().map(file_path_to_string).collect())
    } else {
        builder
            .blocking_pick_file()
            .map(|fp| vec![file_path_to_string(fp)])
    };

    match paths {
        Some(file_paths) => Ok(OpenDialogResult {
            canceled: false,
            file_paths,
        }),
        None => Ok(OpenDialogResult {
            canceled: true,
            file_paths: vec![],
        }),
    }
}

/// 把 Tauri `FilePath`（可能是本地路径或 URL）统一转为字符串。
fn file_path_to_string(fp: FilePath) -> String {
    match fp {
        FilePath::Path(p) => p.to_string_lossy().into_owned(),
        FilePath::Url(u) => u.to_string(),
    }
}
